armory/art_experimental/attacks/carla_obj_det_adversarial_patch.py

```diff
@@ -0,0 +1,598 @@
+
+                if y_preprocessed is not None and isinstance(
+                    y_preprocessed[0]["boxes"], np.ndarray
+                ):
+                    y_preprocessed_tensor = []
+                    for i, y_i in enumerate(y_preprocessed):
+                        y_preprocessed_t = {}
+                        y_preprocessed_t["boxes"] = (
+                            torch.from_numpy(y_i["boxes"])
+                            .type(torch.float)
+                            .to(self.device)
+                        )
+                        y_preprocessed_t["labels"] = (
+                            torch.from_numpy(y_i["labels"])
+                            .type(torch.int64)
+                            .to(self.device)
+                        )
+                        if "masks" in y_i:
+                            y_preprocessed_t["masks"] = (
+                                torch.from_numpy(y_i["masks"])
+                                .type(torch.uint8)
+                                .to(self.device)
+                            )
+                        y_preprocessed_tensor.append(y_preprocessed_t)
+                    y_preprocessed = y_preprocessed_tensor
+
+                transform = torchvision.transforms.Compose(
+                    [torchvision.transforms.ToTensor()]
+                )
+                image_tensor_list_grad = []
+
+                for i in range(x_preprocessed.shape[0]):
+                    if self.clip_values is not None:
+                        x_grad = transform(x_preprocessed[i] / self.clip_values[1]).to(
+                            self.device
+                        )
+                    else:
+                        x_grad = transform(x_preprocessed[i]).to(self.device)
+                    x_grad.requires_grad = True
+                    image_tensor_list_grad.append(x_grad)
+
+                inputs_t = image_tensor_list_grad
+
+            else:
+                raise NotImplementedError(
+                    "Combination of inputs and preprocessing not supported."
+                )
+
+            if isinstance(y_preprocessed, np.ndarray):
+                labels_t = torch.from_numpy(y_preprocessed).to(self.device)  # type: ignore
+            else:
+                labels_t = y_preprocessed  # type: ignore
+
+            output = self._model(inputs_t, labels_t)
+
+            return output, inputs_t, image_tensor_list_grad
+
+    def _train_step(
+        self,
+        images: "torch.Tensor",
+        target: "torch.Tensor",
+        mask: Optional["torch.Tensor"] = None,
+    ) -> "torch.Tensor":
+        import torch  # lgtm [py/repeated-import]
+
+        self.estimator.model.zero_grad()
+        loss = self._loss(images, target, mask)
+        loss.backward(retain_graph=True)
+
+        if self._optimizer_string == "pgd":
+            patch_grads = self._patch.grad
+            if patch_grads.shape[0] == 6:
+                patch_grads[3:, :, :] = torch.mean(
```

Comment:
what's the motivation for averaging over the depth-channel gradients?

Reply:
Depth only has one channel, but we repeat three times for it to work with the model on the forward pass. On the backward pass, the gradients are calculated for three depth channels, but we need to change them into three identical channels. To me, it makes more sense to average the gradients in the three depth channels than to randomly pick gradients for one of the channels.